react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+3. Created a `useEffect` to be executed once on the mount of the component that will try to call the `focus` method of the button element.
+4. We've attached `buttonRef` to the `ref` attribute of the button element. This establishes the connection between the `buttonRef` and the button in the DOM.
+
+Whenever your website loads, it will automatically focus on the button element. You might ask, how can it have the `focus` method when the initial value is `null`? You should by now know that `rendering` and `painting of the screen` comes first before React runs the `useEffect`. It has already established the connection between the ref and the button before the effect is executed.
+
+Also, remember that `useRef` hook isn't just limited to focusing elements. It can be used for various other DOM manipulation scenarios, such as scrolling to a specific position, measuring the dimensions of an element, triggering animations, and basically *any* DOM manipulation that you've done before with vanilla JavaScript. The possibilities are endless! For example, we can change the `useEffect` in the above snippet to do the following. Change the button's text, and after 2 seconds, change the text back. You **should not** do this and only use `useRef` for non-destructive DOM operations, but just an example:
+
+~~~jsx
+useEffect(() => {
+  buttonRef.current.focus();
+  buttonRef.current.textContent = "Hey, I'm different!";
+  let timeout = setTimeout(() => {
+    buttonRef.current.textContent = "Click Me!";
+  }, 2000);
+
+  return () => {
+    clearTimeout(timeout);
+  };
+}, []);
+~~~
+
+The best thing about this is that this will **never** trigger a component re-render!
+
+Another question that might pop up in your mind is, "Why not just use `querySelector` or other DOM manipulation methods that we've done previously in vanilla JavaScript?" Primarily because of performance and using the hook with React, it's ensured that it will be optimized. We should also not stray outside React and use the provided API to work on their ecosystem.
+
+We can also see that it's similar to the `useState` hook in that it can store some values. The main difference is that `useRef` creates a mutable reference, allowing you to update its value without triggering a re-render. But, `useState` manages an immutable state that triggers re-renders when updated.
+
+### The useMemo hook
+
+In all of the examples, I would advise you to use the [Profiler component](https://react.dev/reference/react/Profiler) that is provided in the `react` module. If you want a more interactive alternative, use the `Profiler` in the [React Developer Tools](https://react.dev/learn/react-developer-tools). To measure rendering performance. Note that sometimes you don't need to optimize anything because of how fast things are already. As the famous saying goes in software development:
```

Comment:
```suggestion
In all of the examples, we would advise you to use the [Profiler component](https://react.dev/reference/react/Profiler) that is provided in the `react` module. If you want a more interactive alternative, use the `Profiler` in the [React Developer Tools](https://react.dev/learn/react-developer-tools). To measure rendering performance. Note that sometimes you don't need to optimize anything because of how fast things are already. As the famous saying goes in software development:
```